DRAFT - 22.0.0.1 GA release blog (#1990)
OpenLiberty/blogs: posts/2022-01-18-ga-release-post.adoc

---
layout: post

> title: "MicroProfile 5.0, updates to mpOpenApi-2.0 and many more exciting new features in Open Liberty 22.0.0.1"

Reviewer: ```suggestion
title: "MicroProfile 5.0, JWE support in OIDC and Social Media Login, and many more exciting new features in Open Liberty 22.0.0.1"
```

# Do NOT change the categories section
categories: blog
author_picture: https://avatars3.githubusercontent.com/ryan-storey
author_github: https://github.com/ryan-storey

> seo-title: MicroProfile 5.0, updates to mpOpenApi-2.0 and many more exciting new features in Open Liberty 22.0.0.1 - OpenLiberty.io

Reviewer: ```suggestion
seo-title: MicroProfile 5.0, JWE support in OIDC and Social Media Login, and many more exciting new features in Open Liberty 22.0.0.1 - OpenLiberty.io
```

> seo-description: Open Liberty 22.0.0.1 offers many exciting new features including MicroProfile 5.0, enabling applications to use MicroProfile APIs together with Jakarta EE 9.1. Also included in this update are changes to MicroProfile OpenAPI which suppors multiple applications, as well as many more exciting new changes. This update also comes with several significant bug fixes.

Reviewer: ```suggestion
seo-description: Open Liberty 22.0.0.1 offers many exciting new features including MicroProfile 5.0, enabling applications to use MicroProfile APIs together with Jakarta EE 9.1. Furthermore, the MicroProfile RestClient implementation has switched from Apache CXF to RESTEasy, MicroProfile OpenAPI has been updated to support multiple applications, and JWE support has been added to OpenID Connect Client and Social Media Login. Several other features and many bug fixes are also included in 22.0.0.1, as well as the introduction of signature files (`.sig`) for our downloadable release binaries.
```

> blog_description: Open Liberty 22.0.0.1 offers many exciting new features including MicroProfile 5.0, enabling applications to use MicroProfile APIs together with Jakarta EE 9.1. Also included in this update are changes to MicroProfile OpenAPI which suppors multiple applications, as well as many more exciting new changes. This update also comes with several significant bug fixes.

Reviewer: ```suggestion
blog_description: Open Liberty 22.0.0.1 offers many exciting new features including MicroProfile 5.0, enabling applications to use MicroProfile APIs together with Jakarta EE 9.1. Furthermore, the MicroProfile RestClient implementation has switched from Apache CXF to RESTEasy, MicroProfile OpenAPI has been updated to support multiple applications, and JWE support has been added to OpenID Connect Client and Social Media Login. Several other features and many bug fixes are also included in 22.0.0.1, as well as the introduction of signature files (`.sig`) for our downloadable release binaries.
```

open-graph-image: https://openliberty.io/img/twitter_card.jpg
---

> = MicroProfile 5.0, updates to mpOpenApi-2.0 and many more exciting new features in Open Liberty 22.0.0.1

Reviewer: ```suggestion
= MicroProfile 5.0, JWE support in OIDC and Social Media Login, and many more exciting new features in Open Liberty 22.0.0.1
```

Ryan Storey <https://github.com/ryan-storey>
:imagesdir: /
:url-prefix:
:url-about: /
//Blank line here is necessary before starting the body of the post.

> Open Liberty 22.0.0.1 offers many exciting new features including MicroProfile 5.0, enabling applications to use MicroProfile APIs together with Jakarta EE 9.1. Also included in this update are changes to MicroProfile OpenAPI which suppors multiple applications, as well as many more exciting new changes. This update also comes with several significant bug fixes.

Reviewer: ```suggestion
Open Liberty 22.0.0.1 offers many exciting new features including MicroProfile 5.0, enabling applications to use MicroProfile APIs together with Jakarta EE 9.1. Furthermore, the MicroProfile RestClient implementation has switched from Apache CXF to RESTEasy, MicroProfile OpenAPI has been updated to support multiple applications, and JWE support has been added to OpenID Connect Client and Social Media Login. Several other features and many bug fixes are also included in 22.0.0.1, as well as the introduction of signature files (`.sig`) for our downloadable release binaries.
```

In link:{url-about}[Open Liberty] 22.0.0.1:

* <<mp5, MicroProfile 5.0>>
** <<mpRestClient, MicroProfile Rest Client 3.0 - Jakarta EE 9.1 support>>
* <<mpOpenApi, Support multiple applications in mpOpenApi-2.0>>

> * <<jwe, Support JWE for OpenID Connect Client and Social Media Login>>

Reviewer: ```suggestion
* <<jwe, Support JWE for OpenID Connect Client and Social Media Login>>
* <<block_classes, Block Loading of Classes with Known Vulnerabilities>>
```

* <<release_packages, Release packages are now signed>>
* <<expansion_variables, Using expansion variables in server.env>>
* <<jvm, JVM Working Directory>>
* <<stale_connection_identification, Customizing Stale Connection Identification>>
* <<deployment_descriptor_parsing, Improvements to Deployment Descriptor Parsing>>

> View the list of fixed bugs in link:https://github.com/OpenLiberty/open-liberty/issues?q=label%3Arelease%3A22.0.0.1+label%3A%22release+bug%22[22.0.0.1].

Reviewer: ```suggestion
View the list of fixed bugs in link:https://github.com/OpenLiberty/open-liberty/issues?q=label%3Arelease%3A22001+label%3A%22release+bug%22[22.0.0.1].
```

[#run]
== Run your apps using 22.0.0.1

If you're using link:{url-prefix}/guides/maven-intro.html[Maven], here are the coordinates:

[source,xml]
----
<dependency>
    <groupId>io.openliberty</groupId>
    <artifactId>openliberty-runtime</artifactId>
    <version>22.0.0.1</version>
    <type>zip</type>
</dependency>
----

Or for link:{url-prefix}/guides/gradle-intro.html[Gradle]:

[source,gradle]
----
dependencies {
    libertyRuntime group: 'io.openliberty', name: 'openliberty-runtime', version: '[22.0.0.1,)'
}
----

Or if you're using Docker:

[source]
----
FROM open-liberty
----

// // // // // // // //
// Above:
// Replace the 22.0.0.1 with the Open Liberty release number e.g. 20.0.0.4

> // // // // // // // //

Reviewer: ```suggestion
```

Or take a look at our link:{url-prefix}/downloads/[Downloads page].

[link=https://stackoverflow.com/tags/open-liberty]
image::img/blog/blog_btn_stack.svg[Ask a question on Stack Overflow, align="center"]

[#mp5]
== MicroProfile 5.0

MicroProfile 5.0 enables applications to use MicroProfile APIs together with Jakarta EE 9.1. MicroProfile 5.0 does not provide any other functional updates except aligning with Jakarta EE 9.1.  MicroProfile 5.0 includes Config 3.0, Fault Tolerance 4.0, Rest Client 3.0, Health 4.0, Metrics 4.0, Open Tracing 3.0, Open API 3.0, and JWT propagation 2.0. This is the first release that allows you to use Jakarta namespace together with MicroProfile APIs. 

[#mpRestClient]
=== MicroProfile Rest Client 3.0 - Jakarta EE 9.1 support 

> As a part of this release, Open Liberty consumes RESTEasy for its implementation for MicroProfile Rest Client 3.0 while the previous version of MicroProfile Rest Client uses Apache CXF - this change brings with it some behavior and property changes (most of which are already documented as link:https://draft-openlibertyio.mybluemix.net/docs/22.0.0.1/reference/jaxrs-dif.html[differences] between jaxrs-2.1 and restfulWS-3.0).

Reviewer: ```suggestion
As a part of this release, Open Liberty consumes RESTEasy for its implementation for MicroProfile Rest Client 3.0 while the previous version of MicroProfile Rest Client uses Apache CXF - this change brings with it some behavior and property changes (most of which are already documented as link:{url-prefix}/docs/latest/reference/jaxrs-dif.html[differences] between jaxrs-2.1 and restfulWS-3.0).
```

|`mp.openapi.extensions.liberty.merged.include` | List of modules which should be included in the merged OpenAPI documentation | `first`
a|
* `all` (to include all applications)
* `first` (to include only the first web module deployed, matching the previous behavior)
* comma-separated list of `<appname>` (to include individual applications) and `<appname>/<modulename>` (to include individual modules within an EAR)
|`mp.openapi.extensions.liberty.merged.exclude`
a|
* List of modules which should be excluded from the merged OpenAPI documentation
* Takes priority over the list of included modules
| `none`
a|
* `none` (to exclude nothing)
* comma-separated list of <appname> (to exclude individual applications) and <appname>/<modulename> (to exclude individual modules within an EAR)
|`mp.openapi.extensions.liberty.merged.info`| This property sets the info section of the final Open API document | N/A 
a|
* The value must be a valid OpenAPI info section in JSON format. If this property is set, the info section in the final OpenAPI document is replaced with the value of the property. This replacement is made after any merging is completed.
|===

> For more information, refer to the documentation available link:{url-prefix}/docs/latest/documentation-openapi.html#multi-module[here]

Reviewer: ```suggestion
For more information, refer to the link:{url-prefix}/docs/latest/documentation-openapi.html#multi-module[openAPI multi-module documentation].
```

You can also configure Liberty OIDC Social Login to process ID tokens that are in a `JWE` format. The corresponding OpenID Connect Provider should support creating `JWE` ID tokens.

* Set the OpenID Connect Provider with the OIDC Social Login's public key that is used to encrypt the Content Encryption Key according to the OpenID Connect Provider's documentation.
* Set the `keyManagementKeyAlias` attribute to the private key alias of the key management key that is used to decrypt the Content Encryption Key of JWE token. The key must exist in the keyStore configured for the SSL configuration referred by the sslRef attribute. For example,
+
[source, xml]
----
<oidcLogin keyManagementKeyAlias="privateKeyAlias" />
----

The signatureAlgorithm attributes of both elements now support the `RS384`, `RS512`, `HS384`, `HS512`, `ES256`, `ES384`, and `ES512` signature algorithms.

[source, xml]
----
<openidConnectClient signatureAlgorithm="RS384"/>
<oidcLogin signatureAlgorithm="RS384"/>
----

> 

Reviewer: ```suggestion
[#block_classes] 
== Block Loading of Classes with Known Vulnerabilities (e.g. Apache log4j "JndiLookup")

Applications deployed to Liberty may run versions of Log4j2 that are affected by Log4Shell (CVE-2021-44228) and related vulnerabilities.

This new function modifies the application and library class loaders to block the loading of the `org.apache.logging.log4j.core.lookup.JndiLookup` class, which is the cause of the vulnerability.

Users should analyze their applications for use of Log4j2 with urgency; in the meantime this fix may help mitigate Log4Shell and other vulnerabilities related to the `org.apache.logging.log4j.core.lookup.JndiLookup` class. However, one should note that this will not protect in cases where the Log4j2 classes have been renamed (a process known as "shading") or if Log4j2 is loaded by other class loaders, such the Java system class loader or user-created class loaders. 

```

[#release_packages]
== Release packages are now signed 

Starting in 22.0.0.1 release, we are signing our downloadable binaries. You can use these signature files and the corresponding public key to verify the authenticity and integrity of an Open Liberty release package.

The Open Liberty project uses its private key to digitally sign each Open Liberty release. You can use the Open Liberty public key to check the signature, verify that the package was released by Open Liberty, and that it was not modified since its release.

You can verify a release package either locally, by using the openssl command and a `*.sig` file, or on Maven Central, by using the gpg command.

*Verifying Open Liberty packages with OpenSSL*

To verify an Open Liberty release package locally, you must first download an Open Liberty `.zip` package, the corresponding `.sig` file, and the Open Liberty public key. You can then run the `openssl` command to verify the package.

> Go to the Download package section of the Open Liberty Get Started page and download an Open LIberty `.zip` package and its corresponding `.sig` file.

Reviewer: ```suggestion
Go to the link:{url-prefix}/start/#runtime_releases[Download package section of the Get Started page] and download an Open Liberty `.zip` package and its corresponding `.sig` file.
```

Obtain the public key file by using the public key link on the Get Started page.

After you download the files, you can run the following `OpenSSL` command from the command line to verify the package.

[source]
----
openssl dgst -sha256 -verify WebSphereLiberty_06-02-2021.pem -signature openliberty-javaee8-22.0.0.1.zip.sig openliberty-javaee8-22.0.0.1.zip
This example uses the WebSphereLiberty_06-02-2021.pem public key file and openliberty-javaee8-22.0.0.1.zip.sig signature file to verify the openliberty-javaee8-22.0.0.1.zip release package. Replace the signature file and package version values according to the package that you want to verify. If the verification is successful, the command produces the following console output.

Verified OK

> ----

Reviewer: ```suggestion
After you download the files, use the `openssl` command from the command line to verify the package. For example:

[source]
----
openssl dgst -sha256 -verify WebSphereLiberty_06-02-2021.pem -signature openliberty-javaee8-22.0.0.1.zip.sig openliberty-javaee8-22.0.0.1.zip

Verified OK
----

This example uses the WebSphereLiberty_06-02-2021.pem public key file and openliberty-javaee8-22.0.0.1.zip.sig signature file to verify the openliberty-javaee8-22.0.0.1.zip release package. Replace the signature file and package version values according to the package that you want to verify. If the verification is successful, the command produces the following console output.  
```

*Verify Liberty packages on Maven Central*

To verify Open Liberty packages on Maven Central, you must first download the public key to your local machine by using the `gpg` or `gpg2` command. You must edit the trust level for the key owner. You can then use this key to remotely verify an Open Liberty `.asc` release file on Maven Central.

Run the following command to download the public key file. The key ID value for the public key is `46102B8E`.

[source]
----
gpg2 --keyserver hkp://keyserver.ubuntu.com --recv-keys 46102B8E
Use the gpg2 --edit-key command to set the trust for the key owner to 5. When you run the gpg2 --edit-key command, the console displays information about the key owner, followed by an internal command prompt. At this prompt, enter the trust command. You are then prompted to select a level of trust. Enter 5.

gpg2 --edit-key "WebSphere Liberty"
gpg> trust

Set trust 5

> ----

Reviewer: ```suggestion
Run the following command to download the public key file. The key ID value for the public key is `46102B8E`.

[source]
----
gpg2 --keyserver hkp://keyserver.ubuntu.com --recv-keys 46102B8E
----

Use the `gpg2 --edit-key` command to set the trust for the key owner to 5. When you run the `gpg2 --edit-key` command, the console displays information about the key owner, followed by an internal command prompt. At this prompt, enter the `trust` command. You are then prompted to select a level of trust. Enter `5`.

[source]
----
gpg2 --edit-key "WebSphere Liberty"
gpg> trust

Set trust 5
----
```

Verify the file by running the `gpg2 --verify` command. The following example verifies the Open Liberty 22.0.0.1 release package.

[source]
----
gpg2 --verify openliberty-runtime-22.0.0.1.zip.asc
If the verification is successful, the command produces console output that is similar to the following example.

gpg: assuming signed data in 'openliberty-runtime-22.0.0.1.zip'
gpg: Signature made Wed Nov 24 09:02:44 2021 EST
gpg: using RSA key 91FFD9A642D060B66B802B9D4D210F6946102B8E
gpg: Good signature from "WebSphere Liberty (Works for IBM) contbld@uk.ibm.com" [ultimate]
gpg: aka "WebSphereLiberty" [ultimate]

> ----

Reviewer: ```suggestion
Verify the file by running the `gpg2 --verify` command. The following example verifies the Open Liberty 22.0.0.1 release package:

[source]
----
gpg2 --verify openliberty-runtime-22.0.0.1.zip.asc

gpg: assuming signed data in 'openliberty-runtime-22.0.0.1.zip'
gpg: Signature made Wed Nov 24 09:02:44 2021 EST
gpg: using RSA key 91FFD9A642D060B66B802B9D4D210F6946102B8E
gpg: Good signature from "WebSphere Liberty (Works for IBM) contbld@uk.ibm.com" [ultimate]
gpg: aka "WebSphereLiberty" [ultimate]
----
```

Prior to this update, it might be necessary to provide custom packaged servers with hard-coded values in the server.env. Now server.env can consume these values from the system environment.
On Windows this capability already existed, though it was not documented anywhere.

On Windows, this capability is enabled by default and has always existed. Since this is new to all other operating systems, it is necessary to enable the new capability by adding a comment line near the top of the file:

`# enable_variable_expansion`

Environment variables are specified using `${variable_name}` syntax, except on Windows where the syntax is `!variable_name!`
In the examples below the `LOG_FILE` variable is assigned the value of an environment variable. This example changes the name of the log file from the default name of `console.log`.

Example `server.env` on Windows:

[source]
----
keystore_password=XASEvZMKn3wG6XuTaVYFr8C
LOG_FILE=!CONSOLE_LOG!
----

Example `server.env` on Linux:

[source]
----
# enable_variable_expansion
keystore_password=XASEvZMKn3wG6XuTaVYFr8C
LOG_FILE=${CONSOLE_LOG}
----

> For more information check out the link:https://openliberty.io/docs/21.0.0.11/reference/config/server-configuration-overview.html[Server configuration overview] and link: https://www.ibm.com/docs/en/was-liberty/core?topic=manually-customizing-liberty-environment[Customizing the Liberty environment] documentation.

Reviewer: ```suggestion
For more information check out the link:{url-prefix}/docs/latest/reference/config/server-configuration-overview.html[Server configuration overview] documentation.
```

[#jvm]
== JVM Working Directory

> This enhancement would allow the user to specify the `SERVER_WORKING_DIR` environment variable and set the `JVM` working directory location to something other than the `${WLP_OUTPUT_DIR}/serverName location`. The request from the `RFE` is to NOT make this a fully qualified path for portability purposes. Thus a path that is relative to the `${WLP_OUTPUT_DIR}/serverName` location can be used, as well as, an absolute path(one that contains a `c:\` on Windows, or a `/` on linux based operating systems).

Reviewer: ```suggestion
This enhancement introduces a new `SERVER_WORKING_DIR` environment variable which allows the user to set the `JVM` working directory location to something other than the `${WLP_OUTPUT_DIR}/serverName location`.  For portability purposes, the path supports not only absolute paths (one that contains a `c:\` on Windows, or a `/` on linux based operating systems), but also allows users to specify relative paths to the `${WLP_OUTPUT_DIR}/serverName` directory.
```

> So for example, the user could set the Open Liberty `JVM` output to be added to the `${WLP_OUTPUT_DIR}/serverName/logs` location so that all `JVM` related data would be in with the server log data (relative path example):
> 
> `SERVER_WORKING_DIR=logs`

Reviewer: ```suggestion
For example, the user could set the Open Liberty `JVM` output to be added to the `${WLP_OUTPUT_DIR}/serverName/logs` location so that all `JVM` related data would be in with the server log data (relative path example) by setting the following:

`SERVER_WORKING_DIR=logs`
```

> Or the end user could move the information outside of the `${WLP_OUTPUT_DIR}/serverName` location by doing something similar to the following (which would put the data in the `/wlp/usr/servers/logs/` folder and is an absolute path example for Linux based operating systems):

Reviewer: ```suggestion
Or the user could move the information outside of the `${WLP_OUTPUT_DIR}/serverName` location by doing something similar to the following (which would put the data in the `/wlp/usr/servers/logs/` folder and is an absolute path example for Linux based operating systems):
```

`SERVER_WORKING_DIR=/wlp/usr/servers/logs/`

> An absolute path example on on Windows would look similar to the following:

Reviewer: ```suggestion
An absolute path example on Windows would look similar to the following:
```

[#bugs]
== Notable bugs fixed in this release

> We’ve spent some time fixing bugs. The following sections describe just some of the issues resolved in this release. If you’re interested, here’s the  link:https://github.com/OpenLiberty/open-liberty/issues?q=label%3Arelease%3A22.0.0.1+label%3A%22release+bug%22[full list of bugs fixed in 22.0.0.1].

Reviewer: ```suggestion
We’ve spent some time fixing bugs. The following sections describe just some of the issues resolved in this release. If you’re interested, here’s the link:https://github.com/OpenLiberty/open-liberty/issues?q=label%3Arelease%3A22001+label%3A%22release+bug%22[full list of bugs fixed in 22.0.0.1].
```

*link:https://github.com/OpenLiberty/open-liberty/issues/19631[featureUtility installServerFeature fails when user feature is listed]
+
featureUtility installServerFeatures previously contained bugs when the user configured a product extension in their server.xml, e.g.
+
[source, xml]
----
<feature>myExt:userfeature-1.0</feature>
----
+
If the user feature file doesn't contain the following regex (Liberty version), `\\d\\d\\.\\d\\.\\d\\.\\d\\.esa`, the user feature would not be installed to the defined product extension. The user feature doesn't need to have the same version as the Liberty version.
The tool didn't fetch the installed features properly.
Changes have been made to the underlying code to ensure that the feature name is extracted correctly from the filename.

*link:https://github.com/OpenLiberty/open-liberty/issues/19589[ArrayIndexOutOfBoundsException during startup with mpOpenApi]
+

> During startup, `mpOpenApi` uses `jandex` to index classes and the following exception was observed in link:https://github.com/IBM/FHIR/issues/3092[IBM/FHIR#3092]:

Reviewer: ```suggestion
During startup, `mpOpenApi` uses `jandex` to index classes and the following exception was observed:
```

*link:https://github.com/OpenLiberty/open-liberty/issues/19547[New HTTP/2 streams still accepted while server is closing]
+
Due to an oversight in a previous fix, HTTP/2 streams could still be accepted after the server shutdown process begins.
This would happen during server shutdown, with a quiesce timeout active, and an HTTP/2 connection actively generating new streams. As long as new streams keep the connection open during the quiesce timeout, the quiesce warning message would be invoked. A similar warning would be logged:
+
[source]
----
CWWKE1106W: 1 shutdown operations did not complete during the quiesce period.
----
+
During quiesce, connections should begin closing down. In HTTP/1.1 this implies disabling keep-alive. In HTTP/2, sending a GOAWAY once quiesce begins. We have worked on improving the HTTP/2 server shutdown behaviour to fix this issue.

*link:https://github.com/OpenLiberty/open-liberty/issues/19522[Unresolved gRPC bundles in feature when used alongside servlet-5.0]
+

> Previously, the components within the `grpc` monitor bundles did not properly resolve when running on the EE9 `servlet-5.0` feature. This bug occured during when `gRPC` is running on `servlet-5.0` and some monitoring feature is enabled. For instance, this bug would occur with `grpc-1.0` and `mpMetrics-4.0`. This bug was fixed by updating the range of `io.grpc` versions for EE9.

Reviewer: ```suggestion
Previously, the components within the `grpc` monitor bundles did not properly resolve when running on the Jakarta EE9 `servlet-5.0` feature. This bug occurred when `gRPC` runs alongside the `servlet-5.0` feature and some monitoring feature is enabled. For instance, this bug would occur with `grpc-1.0` and `mpMetrics-4.0`. This bug was fixed by updating the range of `io.grpc` versions for EE9.
```

*link:https://github.com/OpenLiberty/open-liberty/issues/19433[JNDI lookup to CORBA URL can hang]
+

> A naming lookup that results in the first call to the CORBA COSNaming NameService service should cause the service to be activated. However previously, if two concurrent naming lookups found the service not yet activated, a race condition could have occured, and one of the calls could have hung.

Reviewer: ```suggestion
A naming lookup that results in the first call to the CORBA COSNaming NameService should cause the service to be activated. However, previously, if two concurrent naming lookups found the service not yet activated, a race condition could have occurred, and one of the calls could have hung.
```
